File: TrabalhoOO1/src/main/java/poo/interfacefx/t1_poo/ScreenFx.java
package poo.interfacefx.t1_poo;

import javafx.application.Application;
import javafx.beans.property.SimpleObjectProperty;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.geometry.Insets;
import javafx.scene.Node;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.scene.layout.*;
import javafx.scene.text.Text;
import javafx.stage.Stage;
import javafx.scene.image.Image;

import java.util.Objects;

public class ScreenFx extends Application {

    private Stage tela;
    private TableView<Produtos> tabela;
    private static final int LINHAS_POR_PAGINA = 10;
    private TextField nomeInput;
    private TextField descricaoInput;
    private TextField marcaInput;
    private TextField categoriaInput;
    private TextField precoInput;
    private TextField custoInput;
    private TextField buscaInput;
    private ProdutosControlador produtosControlador;
    private Pagination paginacao;

    public static void main(String[] args) {
        launch(args);
    }

    @Override
    public void start(Stage stage) {
        produtosControlador = new ProdutosControlador();

        // Criacao do GridPane.
        GridPane grid = new GridPane();
        grid.setPadding(new Insets(10, 10, 10, 10));
        grid.setVgap(10);
        grid.setHgap(12);

        // Nome do produto.
        Label nomeLabel = new Label("Nome");
        nomeInput = new TextField();
        nomeInput.setPromptText("Nome do produto");

        // Descricao do produto.
        Label descricaoLabel = new Label("Descrição");
        descricaoInput = new TextField();
        descricaoInput.setPromptText("Descrição do produto");

        // Marca do produto.
        Label marcaLabel = new Label("Marca");
        marcaInput = new TextField();
        marcaInput.setPromptText("Marca do produto");

        // Categoria do produto.
        Label categoriaLabel = new Label("Categoria");
        categoriaInput = new TextField();
        categoriaInput.setPromptText("Categoria do produto");

        // Preco do produto.
        Label precoLabel = new Label("Lista de Preço");
        precoInput = new TextField();
        precoInput.setPromptText("Lista de preço produto");

        // Categoria do produto.
        Label custoLabel = new Label("Custo");
        custoInput = new TextField();
        custoInput.setPromptText("Custo do produto");

        // Botoes de criar.
        Button criarBtn = new Button("Criar");
        criarBtn.setOnAction(_ -> {
            produtosControlador.criarProduto(nomeInput, descricaoInput, marcaInput,
                    categoriaInput, precoInput, custoInput, tabela);
            atualizarPaginacao();
        });

        // Filtro de nome.
        buscaInput = new TextField();
        buscaInput.setPromptText("Pesquisar...");
        buscaInput.setPrefWidth(500);
        Button buscaBtn = new Button("Buscar");
        buscaBtn.setOnAction(_ -> atualizarPaginacao());

        // Container do botao buscar.
        HBox buscaContainer = new HBox();
        buscaContainer.setSpacing(10);
        buscaContainer.getChildren().addAll(buscaInput, buscaBtn);
        // Adicionando o botão buscar ao grid.
        GridPane.setConstraints(buscaContainer, 2, 0);

        // Container dos inputs usados pelo usuario.
        VBox inputsContainer = new VBox();
        inputsContainer.setMinWidth(200);
        inputsContainer.setSpacing(15);
        inputsContainer.getChildren().addAll(nomeLabel, nomeInput, descricaoLabel, descricaoInput,
                marcaLabel, marcaInput, categoriaLabel, categoriaInput, precoLabel, precoInput,
                custoLabel, custoInput, criarBtn);
        // Adicionando o container dos inputs ao grid.
        GridPane.setConstraints(inputsContainer, 0, 1);

        // TABELA DE PRODUTOS.

        // Coluna Nomes.
        TableColumn<Produtos, String> nomeCol = getProdutosStringTableColumn("Nome", "nome");

        // Coluna Descricao.
        TableColumn<Produtos, String> descricaoCol = getProdutosStringTableColumn("Descrição", "descricao");

        // Coluna Marca.
        TableColumn<Produtos, String> marcaCol = getProdutosStringTableColumn("Marca", "marca");

        // Coluna Categoria.
        TableColumn<Produtos, String> categoriaCol = getProdutosStringTableColumn("Categoria", "categoria");

        // Coluna de Lista de Preco.
        TableColumn<Produtos, Double> precoCol = new TableColumn<>("Lista de Preço");
        precoCol.setMinWidth(200);
        precoCol.setCellValueFactory(new PropertyValueFactory<>("preco"));

        // Coluna de Custo.
        TableColumn<Produtos, String> custoCol = new TableColumn<>("Custo");
        custoCol.setMinWidth(100);
        custoCol.setCellValueFactory(new PropertyValueFactory<>("custo"));

        // Coluna dos Botoes de Atualizar e Deletar.
        TableColumn<Produtos, BotoesDeOpcoes> opcoesCol = new TableColumn<>("Opções");
        opcoesCol.setMinWidth(175);
        opcoesCol.setCellValueFactory(celldata -> {
            Produtos produto = celldata.getValue();
            return new SimpleObjectProperty<>(new BotoesDeOpcoes(produto, this));
        });

        // Adicionando itens na tabela.
        tabela = new TableView<>();
<<<<<<< HEAD
        tabela.setPrefWidth(776);
        tabela.setItems(produtosControlador.getProdutos());
=======
        tabela.setPrefWidth(885);
        tabela.setItems(getProdutos());
>>>>>>> 6b7c15b2cd10f531cfeed0bf659ec90c09b771d3
        tabela.getColumns().addAll(nomeCol, descricaoCol, marcaCol, categoriaCol, precoCol,
                custoCol, opcoesCol);

        // Paginacao
        paginacao = new Pagination();
        paginacao.setPageFactory(this::criarPagina);
        GridPane.setConstraints(paginacao, 2, 1, 2, 8);

        // Adicionando containers ao grid.
        grid.getChildren().addAll(inputsContainer, buscaContainer, paginacao);

        // Mostrando a cena na tela.
        tela = stage;
        tela.setTitle("Trabalho 1 Poo");
        Scene scene = new Scene(grid, 1280, 720);
        scene.getStylesheets().add(Objects.requireNonNull(getClass().getResource("style.css")).toExternalForm());
        Image icone = new Image(Objects.requireNonNull(getClass().getResource("POO_icon.png")).toExternalForm());
        tela.getIcons().add(icone);
        tela.setScene(scene);
        tela.show();

        // Atualizando a paginacao depois que a tela é inicializada
        atualizarPaginacao();
    }

    public void atualizarProduto(boolean flag, Produtos produto, String novoNome,
                                 String novaDescricao, String novaMarca,
                                 String novaCategoria, double novoPreco, double novoCusto){

        produtosControlador.atualizarProduto(flag, produto, novoNome, novaDescricao,
                novaMarca, novaCategoria, novoPreco, novoCusto, tabela);
        atualizarPaginacao();
    }

    public void deletarProduto(boolean flag, Produtos produto) {
        produtosControlador.deletarProduto(flag, produto, tabela);
        atualizarPaginacao();
    }

    private static TableColumn<Produtos, String> getProdutosStringTableColumn(String Nome, String nome) {
        TableColumn<Produtos, String> auxCol = new TableColumn<>(Nome);
        auxCol.setMinWidth(100);
        auxCol.setCellValueFactory(new PropertyValueFactory<>(nome));
        auxCol.setCellFactory(e -> new TableCell<>() {
            @Override
            protected void updateItem(String item, boolean empty) {
                super.updateItem(item, empty);
                if (item == null || empty) {
                    setText(null);
                    setGraphic(null);
                } else {
                    Text text = new Text(item);
                    text.wrappingWidthProperty().bind
                            (auxCol.widthProperty().subtract(5));
                    setGraphic(text);
                }
            }
        });
        return auxCol;
    }

    private Node criarPagina(int pagina) {
        ObservableList<Produtos> produtosFiltrados = produtosControlador.filtrarProdutos(buscaInput.getText());
        int de = pagina * LINHAS_POR_PAGINA;
        int ate = Math.min(de + LINHAS_POR_PAGINA, produtosFiltrados.size());
        tabela.setItems(FXCollections.observableArrayList(produtosFiltrados.subList(de, ate)));
        return new BorderPane(tabela);
    }

    private void atualizarPaginacao() {
        ObservableList<Produtos> produtosFiltrados = produtosControlador.filtrarProdutos(buscaInput.getText());
        int pageCount = (int) Math.ceil((double) produtosFiltrados.size() / LINHAS_POR_PAGINA);
        paginacao.setPageCount(pageCount);
        paginacao.setPageFactory(this::criarPagina);
    }
}

File: PythonSQL/ScreenFx.py
import tkinter as tk
from tkinter import ttk
from tkinter import messagebox
from functools import partial

class Produtos:
    def __init__(self, nome, descricao, marca, categoria, preco, custo):
        self.nome = nome
        self.descricao = descricao
        self.marca = marca
        self.categoria = categoria
        self.preco = preco
        self.custo = custo

class ProdutosControlador:
    def __init__(self):
        self.produtos = []
    
    def criar_produto(self, nome, descricao, marca, categoria, preco, custo):
        produto = Produtos(nome, descricao, marca, categoria, preco, custo)
        self.produtos.append(produto)
        return produto
    
    def atualizar_produto(self, produto, novo_nome, nova_descricao, nova_marca, nova_categoria, novo_preco, novo_custo):
        produto.nome = novo_nome
        produto.descricao = nova_descricao
        produto.marca = nova_marca
        produto.categoria = nova_categoria
        produto.preco = novo_preco
        produto.custo = novo_custo
    
    def deletar_produto(self, produto):
        self.produtos.remove(produto)
    
    def filtrar_produtos(self, termo_busca):
        return [p for p in self.produtos if termo_busca.lower() in p.nome.lower()]

class ScreenTk:
    def __init__(self, root):
        self.root = root
        self.produtos_controlador = ProdutosControlador()
        
        self.criar_interface()
    
    def criar_interface(self):
        self.root.title("Trabalho 1 Poo")
        
        # Container principal
        main_frame = ttk.Frame(self.root, padding="10 10 10 10")
        main_frame.grid(column=0, row=0, sticky=(tk.N, tk.W, tk.E, tk.S))
        
        # Nome do produto
        ttk.Label(main_frame, text="Nome").grid(column=0, row=0, sticky=tk.W)
        self.nome_input = ttk.Entry(main_frame, width=40)
        self.nome_input.grid(column=0, row=1, padx=5, pady=5, sticky=tk.W)
        
        # Descrição do produto
        ttk.Label(main_frame, text="Descrição").grid(column=0, row=2, sticky=tk.W)
        self.descricao_input = ttk.Entry(main_frame, width=40)
        self.descricao_input.grid(column=0, row=3, padx=5, pady=5, sticky=tk.W)
        
        # Marca do produto
        ttk.Label(main_frame, text="Marca").grid(column=0, row=4, sticky=tk.W)
        self.marca_input = ttk.Entry(main_frame, width=40)
        self.marca_input.grid(column=0, row=5, padx=5, pady=5, sticky=tk.W)
        
        # Categoria do produto
        ttk.Label(main_frame, text="Categoria").grid(column=0, row=6, sticky=tk.W)
        self.categoria_input = ttk.Entry(main_frame, width=40)
        self.categoria_input.grid(column=0, row=7, padx=5, pady=5, sticky=tk.W)
        
        # Preço do produto
        ttk.Label(main_frame, text="Lista de Preço").grid(column=0, row=8, sticky=tk.W)
        self.preco_input = ttk.Entry(main_frame, width=40)
        self.preco_input.grid(column=0, row=9, padx=5, pady=5, sticky=tk.W)
        
        # Custo do produto
        ttk.Label(main_frame, text="Custo").grid(column=0, row=10, sticky=tk.W)
        self.custo_input = ttk.Entry(main_frame, width=40)
        self.custo_input.grid(column=0, row=11, padx=5, pady=5, sticky=tk.W)
        
        # Botão Criar
        criar_btn = ttk.Button(main_frame, text="Criar", command=self.criar_produto)
        criar_btn.grid(column=0, row=12, padx=5, pady=5, sticky=tk.W)
        
        # Container de Busca
        busca_container = ttk.Frame(main_frame, padding="0 10 0 0")
        busca_container.grid(column=0, row=13, padx=5, pady=5, sticky=tk.W)
        
        self.busca_input = ttk.Entry(busca_container, width=50)
        self.busca_input.grid(column=0, row=0, padx=5, pady=5, sticky=tk.W)
        busca_btn = ttk.Button(busca_container, text="Buscar", command=self.atualizar_paginacao)
        busca_btn.grid(column=1, row=0, padx=5, pady=5, sticky=tk.W)
        
        # Tabela de Produtos
        columns = ("Nome", "Descrição", "Marca", "Categoria", "Lista de Preço", "Custo", "Opções")
        self.tabela = ttk.Treeview(main_frame, columns=columns, show="headings")
        for col in columns:
            self.tabela.heading(col, text=col)
            self.tabela.column(col, width=100)
        self.tabela.grid(column=1, row=0, rowspan=10, padx=10, pady=10, sticky=(tk.W, tk.E, tk.N, tk.S))
        
        # Paginação
        self.paginacao = ttk.Treeview(main_frame)
        self.paginacao.grid(column=1, row=10, padx=10, pady=10, sticky=(tk.W, tk.E, tk.N, tk.S))
        
        # Atualizar Paginação
        self.atualizar_paginacao()
        
        self.root.mainloop()
    
    def criar_produto(self):
        nome = self.nome_input.get()
        descricao = self.descricao_input.get()
        marca = self.marca_input.get()
        categoria = self.categoria_input.get()
        preco = float(self.preco_input.get())
        custo = float(self.custo_input.get())
        
        produto = self.produtos_controlador.criar_produto(nome, descricao, marca, categoria, preco, custo)
        self.atualizar_paginacao()
        
        messagebox.showinfo("Produto Criado", f"Produto {produto.nome} criado com sucesso!")
    
    def atualizar_paginacao(self):
        termo_busca = self.busca_input.get().lower()
        produtos_filtrados = self.produtos_controlador.filtrar_produtos(termo_busca)
        
        # Limpar tabela
        for i in self.tabela.get_children():
            self.tabela.delete(i)
        
        # Preencher tabela
        for produto in produtos_filtrados:
            self.tabela.insert("", "end", values=(produto.nome, produto.descricao, produto.marca,
                                                  produto.categoria, produto.preco, produto.custo, "Editar/Deletar"))
        
        # Atualizar paginação
        # (implementar paginação manualmente no tkinter é mais complexo, isso pode variar de acordo com a necessidade)
        
if __name__ == "__main__":
    root = tk.Tk()
    app = ScreenTk(root)